### ISSUE TITLE
How parse GO types from message bug

### ISSUE THREAD

**Issue by Xeue** [go-gst/go-gst#121]
Hey, we are trying to use the level element which produces messages as described here:
https://gstreamer.freedesktop.org/documentation/level/index.html?gi-language=c

There is no parser for this built in and we are struggling to work out how to access the GValueArrays.
The [GstClockTime] types are converted but the GetValues and Values functions. But the [GValueArray of doubles] is serialised.

We have tried messing around with go glib and havn't really got anywhere other than a few SIGSEGV violations and other such nasty things...

Is there a way to parse these values and if not how can we go about writing a parser?

Thanks!

**Comment by RSWilli** [comment]
The "parsers" are actually called marshalers and are defined here: https://github.com/go-gst/go-glib/blob/main/glib/gvalue.go#L326 in `go-glib`. Additional marshalers can be registered by other libraries like this: https://github.com/go-gst/go-gst/blob/main/gst/gstwebrtc/session_description.go#L13-L20

The two functions `ToGValue` and `marshalXXX` work as follows:

* `ToGValue` gets called when the values need to get passed into cgo, it creates a `GValue` that holds the value and a type info
* `marshalXXX` gets registered for the specific type info and gets called when `GValue.GoValue()` is called.

-----

Since the `GValueArray` is located [in glib](https://docs.gtk.org/gobject/struct.ValueArray.html), the bindings should be added to `go-glib`.

Oftentimes the [gtk rust bindings](https://github.com/gtk-rs/gtk-rs-core/blob/76382518ec9acbdca0735949ce9dffdd5301b1bf/glib/src/value_array.rs#L255) are also a good place for reference, since they do similar things.

**Comment by RSWilli** [comment]
@Xeue did you solve your problem?

**Comment by Xeue** [comment]
Yes, eventually, was a bit long winded, but got what I needed working!

**Comment by grf1** [comment]
hi @Xeue I'm having a similar issue, would you mind sharing what you did to get it working?

**Comment by Xeue** [comment]
Hi grf1, so in this case I was trying to process the a message from the event bus, that message was a message generated in the C code in gstreamer, so is stucted and has types in C. To be able to access that in go we need to "un-marshal" it from a C object and into a GO one.
To do this we use some of the go and C integration methods. It helps reading up on cGo: https://dev.to/metal3d/understand-how-to-use-c-libraries-in-go-with-cgo-3dbn

But essentialy the steps here are the create a function that can unmarshal the C object. A simple versions looks like this:
https://github.com/go-gst/go-gst/blob/main/gst/gstwebrtc/session_description.go#L111-L120

Notice at the top of the file C has been imported, this allows line 112 to use a C function C.g_value_get_boxed (see cGo refernece above)

Once you have a function that can un-marshal it from C you can then register that as an unmarshaler with glib:
https://github.com/go-gst/go-gst/blob/main/gst/gstwebrtc/session_description.go#L15-L19

Once regsitered you can use the usual methods in this library to read the values in go (GValue.GoValue())

I know I havn't explained that particularly well... But hopefully that helps get you going in the direction you need?

**Comment by RSWilli** [comment]
@Xeue could you create a PR for this?

**Comment by Xeue** [comment]
Sure! I'll need to tidy stuff up and remove the stuff that's specific to me, so won't be for a little bit as I've got a lot on at the moment

**Comment by mwesigwadi** [comment]
Hello @Xeue  @RSWilli  could you please provide an example on this I am facing a realted issue from srtsink element. I am trying to read the stats prop form the element which produces gst.Structure nested under callers whhich is a GvaluArray. I have tried a way out but don't understand exactly how these marshaller's work and how to pass the data from c to go.

**Comment by RSWilli** [comment]
@mwesigwadi essentially the marshalers are a way to unpack types from a GValue. GValue is glib's generic value container. You can think of it as a golang `any`, and the marshalers as the casting functions. A marshaler gets registered for a specific type and gets called when we call `glib.Value.GoValue()`. Generally this function should take the C pointer/value and wrap it in a go struct or type, so that none of the unsafe C stuff is exposed.

For GValueArray this should be a struct, which then can implement the methods needed to extract a single value.

I don't have an example for GValueArrays, but for other types there are plenty in go-glib/go-gst: https://github.com/go-gst/go-gst/blob/main/gst/gstwebrtc/session_description.go#L13-L24

**Comment by RSWilli** [comment]
this works now, as of #170 and #216
